### PR TITLE
[Swift] Use ISOFullDate for 'date' format

### PR DESCRIPTION
**Pull request by jgavris**
### PR checklist

- [ x ] Read the [contribution guildelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [ x ] Ran the shell/batch script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run`./bin/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates)
- [ x ] Filed the PR against the correct branch: master for non-breaking changes and `2.3.0` branch for breaking (non-backward compatible) changes.

### Description of the PR

- Adds support for the `date` format (ISO-8601 RFC 3339). A common use case is 'date of birth', where we do not care about units beyond the 'day'.

### CODE REVIEW

**Comment by jgavris** [comment]
cc @vineetshah

**Comment by wing328** [comment]
@jgavris thanks for the PR.

cc @Edubits @jaz-ah

**Comment by jaz-ah** [comment]
@jgavris just out of curiosity, why not have the generator still deal w/ basic objects (NSDate) and then you have convertors in your proj that convert to different ISO formats - i.e. what if other people wanted other ISO formats out of the NSDate...

**Comment by jgavris** [comment]
@jaz-ah The swagger spec states that the `date` format represents an ISO-8601 `full-date`, which is unambiguously `YYYY-MM-DD`. That is the format that I've implemented here. I'm not sure what you mean when you say people might want other formats out of the `NSDate`. If you've specified `type=string,format=date` in your API, you should get a `full-date` field.

**Comment by jaz-ah** [comment]
@jgavris agreed that the server will send down an ISO-8601, but the most convenient/extendable class to deal w/ this in the client is a NSDate - i.e. you can deserialize the ISO-8601 and return an NSDate/serialize NSDate and serialize as ISO-8601 format...

**Comment by jgavris** [comment]
@jaz-ah I agree with you in principle, but that's the implementation problem. The current client cannot do that, and that's how I ended up here. I was implementing a 'date' format field and on my first request, the server rightfully rejected it as being in the wrong format. The Swift codegen encodes all types to a json string using the following protocol.

```swift
protocol JSONEncodable {
    func encodeToJSON() -> AnyObject
}
```

`NSDate` in particular is encoded into the full ISO-8601 `date-time` format with the following

```swift
private let dateFormatter: NSDateFormatter = {
    let fmt = NSDateFormatter()
    fmt.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSZZZZZ"
    fmt.locale = NSLocale(localeIdentifier: "en_US_POSIX")
    return fmt
}()

extension NSDate: JSONEncodable {
    func encodeToJSON() -> AnyObject {
        return dateFormatter.stringFromDate(self)
    }
}
```

The client only knows it has an `NSDate` field, and does not know what format to serialize to, since this converter function does not take any arguments. Without major (and potentially undesirable) changes in the architecture of the Swift codegen, using a specific type is the way to implement this. 

I've provided an API that is lighter than `NSDateComponents` (because I wrap it) so this is effectively doing the same thing in a more friendly way.

**Comment by jaz-ah** [comment]
@jgavris fair enough - looks good @wing328 let's merge.

File: modules/swagger-codegen/src/test/resources/2_0/petstore.json [hunk 1]
```diff
@@ -833,6 +833,10 @@
         "username": {
           "type": "string"
         },
+        "dateOfBirth": {
+          "type": "string",
```
Comment (cbornet) [inline modules/swagger-codegen/src/test/resources/2_0/petstore.json]: Is it OK to modify the petstore.json ?

Comment (wing328) [inline modules/swagger-codegen/src/test/resources/2_0/petstore.json]: Agreed with @cbornet that we should keep Petstore.json as it's.

@jgavris what I would suggest is to add the test case in [petstore-with-fake-endpoints-models-for-testing.yaml](https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/test/resources/2_0/petstore-with-fake-endpoints-models-for-testing.yaml#L989) instead and then update `swift-petstore.sh` to use `petstore-with-fake-endpoints-models-for-testing.yaml`, which covers a lot more edge cases (you may want to do this in another PR as the edge cases in `petstore-with-fake-endpoints-models-for-testing.yaml` may require some fixes in the Swift generators or templates.

Comment (jgavris) [inline modules/swagger-codegen/src/test/resources/2_0/petstore.json]: Thanks! I haven't been able to make the changes yet, but I will.

**Comment by wing328** [comment]
@jgavris I've filed #3945 to remove `dateOfBirth` and therefore I'm closing this PR. Thanks again for the contribution.

(the commit/change is still owned by you)

**Comment by ahmettaha** [comment]
Hello,

Is it possible to make the 

`fmt.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSZZZZZ"`

customizable. I would like to send local date to our server. 

Thank you.

**Comment by jgavris** [comment]
@ahmettaha Could you open an issue with your question specifically? This pull request was to add support for 'dates' specifically (where you want to express a single day).

**Comment by ahmettaha** [comment]
i did. thank you. it's issue

#5318
